File: hosts/ios/App/Source/Gateway/KeychainPrimitives.swift
import Foundation
import Security

/// keychainGet / keychainSet (contract/primitives.md §4) over SecItem
/// generic-password items: service = bundle id, account = the opaque KeyRef.
/// An unset ref reads as null; `keychainSet(ref, null)` deletes; a set on a
/// missing ref adds, on an existing ref updates. iOS is never "unavailable"
/// here — the descriptor declares keychain* available (conformance §7).
final class KeychainPrimitives {
    private let service: String

    init(core: GatewayCore) {
        service = Bundle.main.bundleIdentifier ?? "org.dsh.DSHSpike"
        core.register(name: "keychainGet") { call, done in self.get(call, done) }
        core.register(name: "keychainSet") { call, done in self.set(call, done) }
    }

    private func query(_ ref: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: ref,
        ]
    }

    private func get(_ call: GatewayCall, _ done: @escaping GatewayDone) {
        guard let ref = call.string("ref"), !ref.isEmpty else {
            return done(.failure(Self.invalid("keychainGet", "empty ref")))
        }
        var query = self.query(ref)
        query[kSecReturnData as String] = true
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecItemNotFound { return done(.success(NSNull())) }
        guard status == errSecSuccess, let data = item as? Data else {
            return done(.failure(Self.io("keychainGet", status)))
        }
        done(.success(["secretB64": data.base64EncodedString()]))
    }

    private func set(_ call: GatewayCall, _ done: @escaping GatewayDone) {
        guard let ref = call.string("ref"), !ref.isEmpty else {
            return done(.failure(Self.invalid("keychainSet", "empty ref")))
        }
        let secret = call.args["secretB64"]
        if secret == nil || secret is NSNull {
            SecItemDelete(query(ref) as CFDictionary)
            return done(.success(NSNull()))
        }
        guard let data = secret as? String, let bytes = Data(base64Encoded: data) else {
            return done(.failure(Self.invalid("keychainSet", "malformed secretB64")))
        }
        store(ref, bytes: bytes, done)
    }

    private func store(
        _ ref: String, bytes: Data, _ done: @escaping GatewayDone
    ) {
        let update = [kSecValueData as String: bytes]
        let status = SecItemUpdate(query(ref) as CFDictionary, update as CFDictionary)
        if status == errSecSuccess { return done(.success(NSNull())) }
        guard status == errSecItemNotFound else {
            return done(.failure(Self.io("keychainSet", status)))
        }
        var add = query(ref)
        add[kSecValueData as String] = bytes
        let addStatus = SecItemAdd(add as CFDictionary, nil)
        guard addStatus == errSecSuccess else {
            return done(.failure(Self.io("keychainSet", addStatus)))
        }
        done(.success(NSNull()))
    }

    private static func invalid(_ primitive: String, _ message: String) -> GatewayError {
        GatewayError(code: "invalid", primitive: primitive, message: message)
    }

    private static func io(_ primitive: String, _ status: OSStatus) -> GatewayError {
        GatewayError(code: "io", primitive: primitive,
                     message: "SecItem status \(status)")
    }
}

File: hosts/android/app/src/main/java/com/dshmobile/spike/KeychainPrimitives.kt
package com.dshmobile.spike

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import org.json.JSONObject
import java.io.File
import java.security.KeyStore
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * keychainGet / keychainSet (contract/primitives.md §4) over the Android
 * Keystore — the platform analogue of hosts/ios KeychainPrimitives.swift's
 * SecItem generic-password items. AndroidKeyStore stores KEYS, not blobs, so
 * a secret is sealed with a hardware-backed AES-256-GCM key (created once
 * under the fixed alias) and the ciphertext persists in the app-private
 * `<filesDir>/keychain/<sha256(ref)>` — service = application id, account =
 * the opaque KeyRef, same trust level as iOS' item lookup. An unset ref
 * reads as null; `keychainSet(ref, null)` deletes; set on a missing ref adds,
 * on an existing ref overwrites. The descriptor declares keychain* available
 * — never faked (conformance §7).
 */
class KeychainPrimitives(private val context: Context) {

    companion object {
        private const val KEY_ALIAS = "dsh-keychain-master"
        private const val GCM_TAG_BITS = 128

        private fun invalid(primitive: String, message: String) =
            GatewayCore.GatewayError("invalid", primitive, message)

        private fun io(primitive: String, message: String) =
            GatewayCore.GatewayError("io", primitive, message)

        private fun b64(bytes: ByteArray): String =
            java.util.Base64.getEncoder().encodeToString(bytes)
    }

    private val dir = File(context.filesDir, "keychain").apply { mkdirs() }

    fun register(on: GatewayCore) {
        on.register("keychainGet") { call, done -> get(call, done) }
        on.register("keychainSet") { call, done -> set(call, done) }
    }

    private fun fileFor(ref: String): File {
        val name = MessageDigest.getInstance("SHA-256").digest(ref.toByteArray())
            .joinToString("") { "%02x".format(it) }
        return File(dir, name)
    }

    private fun get(call: GatewayCore.GatewayCall, done: GatewayCore.Done) {
        val ref = call.string("ref")
        if (ref.isNullOrEmpty()) return done.settle(null, invalid("keychainGet", "empty ref"))
        try {
            val file = fileFor(ref)
            if (!file.isFile) return done.settle(null, null) // unset ref reads null
            val blob = file.readBytes()
            val iv = blob.copyOfRange(0, 12)
            val ciphertext = blob.copyOfRange(12, blob.size)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, masterKey(), GCMParameterSpec(GCM_TAG_BITS, iv))
            done.settle(JSONObject().put("secretB64", b64(cipher.doFinal(ciphertext))), null)
        } catch (e: Exception) {
            done.settle(null, io("keychainGet", "${e::class.java.simpleName}: ${e.message}"))
        }
    }

    private fun set(call: GatewayCore.GatewayCall, done: GatewayCore.Done) {
        val ref = call.string("ref")
        if (ref.isNullOrEmpty()) return done.settle(null, invalid("keychainSet", "empty ref"))
        try {
            if (call.args.isNull("secretB64")) {
                fileFor(ref).delete() // keychainSet(ref, null) deletes
                return done.settle(null, null)
            }
            val bytes = java.util.Base64.getDecoder().decode(call.args.getString("secretB64"))
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, masterKey())
            val blob = cipher.iv + cipher.doFinal(bytes)
            fileFor(ref).writeBytes(blob)
            done.settle(null, null)
        } catch (e: Exception) {
            done.settle(null, io("keychainSet", "${e::class.java.simpleName}: ${e.message}"))
        }
    }

    /** The fixed-alias AES key: created on first use, never extractable. */
    private fun masterKey(): SecretKey {
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        (ks.getKey(KEY_ALIAS, null) as? SecretKey)?.let { return it }
        val generator = KeyGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore",
        )
        generator.init(
            KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build(),
        )
        return generator.generateKey()
    }
}
